runners/spark/src/main/java/org/apache/beam/runners/spark/translation/SparkInputDataProcessor.java

```diff
@@ -0,0 +1,371 @@
+import org.checkerframework.checker.nullness.qual.Nullable;
+import scala.Tuple2;
+
+/**
+ * Processes Spark's input data iterators using Beam's {@link
+ * org.apache.beam.runners.core.DoFnRunner}.
+ */
+interface SparkInputDataProcessor<FnInputT, FnOutputT, OutputT> {
+
+  /**
+   * @return {@link OutputManager} to be used by {@link org.apache.beam.runners.core.DoFnRunner} for
+   *     emitting processing results
+   */
+  OutputManager getOutputManager();
+
+  /**
+   * Processes input partition data and return results as {@link Iterable}.
+   *
+   * @param input input partition iterator
+   * @param ctx current processing context
+   */
+  <K> Iterator<OutputT> process(
+      Iterator<WindowedValue<FnInputT>> input, SparkProcessContext<K, FnInputT, FnOutputT> ctx);
+
+  /**
+   * Creates {@link SparkInputDataProcessor} which does processing in calling thread. It is doing so
+   * by processing input element completely and then iterating over the output retrieved from that
+   * processing. The result of processing one element must fit into memory.
+   */
+  static <FnInputT, FnOutputT>
+      SparkInputDataProcessor<FnInputT, FnOutputT, Tuple2<TupleTag<?>, WindowedValue<?>>>
+          createUnbounded() {
+    return new UnboundedSparkInputDataProcessor<>();
+  }
+
+  /**
+   * Creates {@link SparkInputDataProcessor} which does process input elements in separate thread
+   * and observes produced outputs via bounded queue in other thread. This does not require results
+   * of processing one element to fit into the memory.
+   */
+  static <FnInputT, FnOutputT>
+      SparkInputDataProcessor<FnInputT, FnOutputT, Tuple2<TupleTag<?>, WindowedValue<?>>>
+          createBounded() {
+    return new BoundedSparkInputDataProcessor<>();
+  }
+}
+
+class UnboundedSparkInputDataProcessor<FnInputT, FnOutputT>
+    implements SparkInputDataProcessor<FnInputT, FnOutputT, Tuple2<TupleTag<?>, WindowedValue<?>>> {
+
+  private final UnboundedDoFnOutputManager outputManager = new UnboundedDoFnOutputManager();
+
+  @Override
+  public OutputManager getOutputManager() {
+    return outputManager;
+  }
+
+  @Override
+  public <K> Iterator<Tuple2<TupleTag<?>, WindowedValue<?>>> process(
+      Iterator<WindowedValue<FnInputT>> input, SparkProcessContext<K, FnInputT, FnOutputT> ctx) {
+    return new UnboundedInOutIterator<>(input, ctx);
+  }
+
+  private static class UnboundedDoFnOutputManager
+      implements OutputManager, Iterable<Tuple2<TupleTag<?>, WindowedValue<?>>> {
+
+    private final ArrayDeque<Tuple2<TupleTag<?>, WindowedValue<?>>> outputs = new ArrayDeque<>();
+
+    public void clear() {
+      outputs.clear();
+    }
+
+    @Override
+    public Iterator<Tuple2<TupleTag<?>, WindowedValue<?>>> iterator() {
+      return outputs.iterator();
+    }
+
+    @Override
+    public synchronized <T> void output(TupleTag<T> tag, WindowedValue<T> output) {
+      outputs.push(new Tuple2<>(tag, output));
```

Comment:
`push` corresponds to `addFirst`, let's use `add` / `addLast` to avoid reversing the order here

Reply:
Fixed